hasktorch/src/Torch/Distributions/Distribution.hs

```diff
@@ -0,0 +1,76 @@
+module Torch.Distributions.Distribution (
+    Scale,
+    Distribution(..),
+    stddev,
+    perplexity,
+    logits_to_probs,
+    clamp_probs,
+    probs_to_logits,
+    extended_shape,
+) where
+
+import Torch.Typed.Functional
+import Torch.Functional.Internal (binary_cross_entropy_with_logits, log)
+import Torch.TensorFactories (ones, onesLike)
+import Torch.TensorOptions
+import qualified Torch.Tensor as D
+import qualified Torch.DType as D
+import qualified Torch.TensorFactories as D
+import qualified Torch.Functional as F
+import Torch.Scalar
+import Torch.Distributions.Constraints
+
+data Scale = Probs | Logits
+
+class Distribution a where
+    batch_shape :: a -> [Int]
+    event_shape :: a -> [Int]
+    expand :: a -> [Int] -> a
+    support :: a -> Constraint
+    mean :: a -> D.Tensor
+    variance :: a -> D.Tensor
+    sample :: a -> [Int] -> IO D.Tensor
+    log_prob :: a -> D.Tensor -> D.Tensor
+    entropy :: a -> D.Tensor
+    enumerate_support :: a -> Bool -> D.Tensor -- (expand=True)
```

Comment:
Isn't it used to output tensor?  Or isn't statistics-package usefull?
(monad-bayes is based on it, so I want to know why not use it.)
For example, statistics's mean is below.
```
class Distribution d => MaybeMean d where
  maybeMean :: d -> Maybe Double
```
http://hackage.haskell.org/package/statistics-0.15.2.0/docs/Statistics-Distribution.html

Reply:
I actually was trying out an external package, `mwc-random`, in turn using `vector`; I kinda felt like the type conversions were complicating things and it'd be nice if we had a tensor-based version.
I based this on PyTorch, and I guess some of these operations may yield gradients as well, though disclaimer I'm not yet sure of that's useful in any way (and in fact I know `sample` explicitly disabled gradients there -- wasn't sure how to translate that here).

Reply:
so `mean` / `variance` return tensors now. I think potentially parts of these could end up as real numbers, with other parts as NaN instead. Under that presumption I think returning `Tensor` (as done in PyTorch) may make sense.